### DISCUSSION TITLE
Get Framed-IPv6-Address from radius

### DISCUSSION THREAD

**Discussion by pedda4711** [strongswan/strongswan#1644] [Q&A]
I've setup strongswan with ikeV2 eap-radius and also with ikev1 l2tp ( bgp routing and different ipsec and openvpn servers, different group rights from AD handelt by freeradius  )
Works well with getting ipv4 from radius.  

My ip-pools from radius also send Framed-IPv6 addresses, but which is the right option to get ipV6 from radius
for rightsourceip? 

rightsourceip=%radius
is there something like %radius6 ?
I get only a ipV6 if i define a pool in rightsourceip

Greetings 

Peter

**Comment by tobiasbrunner** [comment]
This depends on the client and whether it requests a virtual IPv6 address. Dual family with IKEv1 is tricky though, as that requires multiple CHILD_SAs (one for each family). But your setup sounds horrible anyway (don't use IKEv1, don't use L2TP). If you are interested in doing anything route-based, have a look at [this](https://docs.strongswan.org/docs/5.9/features/routeBasedVpn.html).

**Comment by pedda4711** [comment]
Hello Tobias, 

my set up is not horrible. Please ignore the ikev1/l2tp part. 
The xl2tpd is patched to greatly to use ipv6 . I use internal bgp-Routing for a  large (and old growing) network. This parts works very well for the last 10 years, and now also with strongswan  (after find out how to import Kernel-Routingtable 220 to bgp in ffr routing)

I want to migrate to ikeV2 EAP Radius.

The Client (in my tests a Vodafone Handy and a Windows-Client with public ipV6 Adress connect to strongswan and get a ipV6 only from
the ipV6 Pool set in rightsourceip, not from radius. The ipv4 from radius announced to the clients and work well.
But how to announce the ipv6 from radius to client?

**Reply by tobiasbrunner** [reply]
> Please ignore the ikev1/l2tp part.

Then why mention it?





> But how to announce the ipv6 from radius to client?

Works perfectly fine in my tests using FreeRADIUS and assigning both a `Framed-IP-Address` and a `Framed-IPv6-Address` if the client requests an IPv4 and an IPv6 address. Check the logs for what happens in your case.

**Reply by pedda4711** [reply]
> Then why mention it?
i only describe a part of my setup

> Works perfectly fine in my tests using FreeRADIUS and assigning both a `Framed-IP-Address` and a `Framed-IPv6-Address` if the client requests an IPv4 and an IPv6 address. Check the logs for what happens in your case.

Is there a described test scenario where i can take a look?
I'm testing with strongSwan 5.9.10

if a Windows-11 client connect  with rightsourceip=%radius,fd8a:a137:6a67:da69::/97 (leftsubnet=0.0.0.0/0,::/0)
Client get both. A ipV4 from radius and an ipV6 from strongSwan 
The strongSwan ipV6 is announced to (and not from) radius. I can see this in radius logs
 
Apr  4 18:25:15 alwin strongswan[48672]: 10[KNL] adding policy fd8a:a137:6a67:da69::1/128 === ::/0 in [priority 334463, refcount 1]
Apr  4 18:25:15 alwin strongswan[48672]: 10[KNL] adding policy fd8a:a137:6a67:da69::1/128 === ::/0 fwd [priority 334463, refcount 1]
Apr  4 18:25:15 alwin strongswan[48672]: 10[KNL] adding policy ::/0 === fd8a:a137:6a67:da69::1/128 out [priority 334463, refcount 1]
Apr  4 18:25:15 alwin strongswan[48672]: 10[KNL] getting a local address in traffic selector ::/0
Apr  4 18:25:15 alwin strongswan[48672]: 10[KNL] using host %any6

if i set to rightsourceip to %radius only, the client get only the ipv4 from radius 

if have set the radius-pool to 

ippool routing_pool {
        name="routing_pool"
        range_start = 10.37.175.10
        range_stop = 10.37.175.254
        netbits = 24
        ipv6-range-start = fd8a:a137:6a75:da69:0000:0000:0000:0000
        ipv6-range-stop = fd8a:a137:6a75:da69:0000:0000:7fff:ffff
        ipv6-netbits = 97
        filename = ${db_dir}/db_routing.ippool
        ip_index = ${db_dir}/db_routing.ipindex
        override = yes
        maximum_timeout = 0
        key = "%{NAS-IP-Address} %{NAS-Port}"
}

The ipv4 comes from this pool.

**Reply by tobiasbrunner** [reply]
> Is there a described test scenario where i can take a look?

I just changed the [ikev2/rw-eap-framed-ip-radius](https://strongswan.org/testing/testresults/ikev2/rw-eap-framed-ip-radius/) test scenario a bit.

<details><summary>Changes in scenario</summary>

```diff
diff --git a/testing/tests/ikev2/rw-eap-framed-ip-radius/hosts/alice/etc/freeradius/3.0/users b/testing/tests/ikev2/rw-eap-framed-ip-radius/hosts/alice/etc/freeradius/3.0/users
index ba92f0080e23..ad4c6bbd03a3 100644
--- a/testing/tests/ikev2/rw-eap-framed-ip-radius/hosts/alice/etc/freeradius/3.0/users
+++ b/testing/tests/ikev2/rw-eap-framed-ip-radius/hosts/alice/etc/freeradius/3.0/users
@@ -1,4 +1,5 @@
 carol	Cleartext-Password := "Ar3etTnp"
-		Framed-IP-Address = 10.3.0.1
+		Framed-IP-Address = 10.3.0.1,
+		Framed-IPv6-Address = fec4::1
 dave	Cleartext-Password := "W7R0g3do"
 		Framed-IP-Address = 10.3.0.2
diff --git a/testing/tests/ikev2/rw-eap-framed-ip-radius/hosts/carol/etc/swanctl/swanctl.conf b/testing/tests/ikev2/rw-eap-framed-ip-radius/hosts/carol/etc/swanctl/swanctl.conf
index 6dda22241b8e..77982b63c610 100755
--- a/testing/tests/ikev2/rw-eap-framed-ip-radius/hosts/carol/etc/swanctl/swanctl.conf
+++ b/testing/tests/ikev2/rw-eap-framed-ip-radius/hosts/carol/etc/swanctl/swanctl.conf
@@ -3,7 +3,7 @@ connections {
    home {
       local_addrs  = 192.168.0.100
       remote_addrs = 192.168.0.1
-      vips = 0.0.0.0
+      vips = 0.0.0.0, ::
 
       local {
          auth = eap
@@ -15,7 +15,7 @@ connections {
       }
       children {
          home {
-            remote_ts = 10.1.0.0/16
+            remote_ts = 10.1.0.0/16, fec1::/16
 
             updown = /usr/local/libexec/ipsec/_updown iptables
             esp_proposals = aes128gcm128-x25519
``` 

</details> 

> I can see this in radius logs

Which has absolutely nothing to do with RADIUS. What you see there is the installation of some of the IPsec policies of a CHILD_SA in the kernel.

> if i set to rightsourceip to %radius only, the client get only the ipv4 from radius

Then most likely the RADIUS server does not send an IPv6 address.

> if have set the radius-pool to

Hm, is that FreeRADIUS? Because I don't see any `ipv6-*` options defined for `rlm_ippool` [here](https://networkradius.com/doc/current/raddb/mods-available/ippool.html).

Also note that if you use accounting to release the IPs (i.e. via Accounting-Stop) the key you configured isn't ideal because `NAS-Port` will change when IKE_SAs are rekeyed. For a stable ID use `Acct-Session-Id`, see [the _eap-radius_ docs](https://docs.strongswan.org/docs/5.9/plugins/eap-radius.html#_attributes_sent_to_radius_servers) for details.

**Reply by pedda4711** [reply]
Thank you Tobias.  

could you put this frr bgp example to your docs? It imports routing table 220. So if a client connect the route go up and propagate via bgp
I've searched a week for this answer and it helps to announce the routes without any modification on strongswan 

```
frr version 8.4.2
frr defaults traditional
hostname x.y.z
log syslog informational
ip import-table 220
!
debug bgp keepalives
debug bgp updates in
debug bgp updates out
!
router bgp 4711
 bgp router-id 10.10.10.10
 no bgp ebgp-requires-policy
 neighbor 10.10.10.11 remote-as 4712
 neighbor 10.10..10.11 password PASSWORD
 !
 address-family ipv4 unicast
  network 10.10.0.0/16
  network 192.168.0.0/16
  redistribute kernel
  redistribute connected
  redistribute table 220
  neighbor 10.10.10.11 prefix-list TESTSERVER in
  neighbor 10.10.10.11 prefix-list TESTSERVER out
  table-map kernel-in
 exit-address-family
exit
!
ip prefix-list TESTSERVER seq 3 permit 10.10.0.0/16 le 32
ip prefix-list TESTSERVER seq 5 permit 192.168.0.0/16 le 32
!
route-map kernel-in permit 220
exit
```

**Reply by tobiasbrunner** [reply]
I assume Google etc. will eventually find it here as well. I currently don't see where in the docs it could fit (maybe [here](https://docs.strongswan.org/docs/5.9/howtos/forwarding.html) somewhere). And it looks like a lot of these settings are very specific to your setup, so it would probably be a good idea to just point out the relevant options (I guess `ip import-table 220` and `redistribute table 220`?) and explain what they do and where to add them. You could also create a new topic (with proper title etc.) in the ["Show and tell" section](https://github.com/strongswan/strongswan/discussions/categories/show-and-tell) of this forum.

**Reply by pedda4711** [reply]
Routing based VPN in your Documentation is the right place.
As Linux BGP-Routing hint with frr Daemon 

This is a very simple bgp routing sample which use the default behavior from strongswan to use vpn routes in (linux) Kernel table 220 (xfrm)
The goal is to import the route immediately into bgp routing , if a vpn comes up,
The route is also visible and announced to standard-Routing Table. So you can see it also with netstat 
There is no need to write any other updown-scripts 

Following Options are relevant:

**ip import-table 220** 

address-family ipv4 unicast
 ..
 **table-map kernel-in**
exit-address-family
exit

**route-map kernel-in permit 220**

**Reply by tobiasbrunner** [reply]
> Routing based VPN in your Documentation is the right place.

Hm, I don't really think so. This has nothing specifically do to with route-based VPNs. It's still a regular policy-based VPN (it could be a route-based setup, but it does not rely on it - and in most of them no routes are installed in table 220 anyway). As far as I understand, you just let hosts behind your gateway know that it is responsible for specific client IPs (i.e. automating the forwarding via your VPN gateway instead of e.g. configuring or announcing a static route to the complete virtual IP subnet as described [on the forwarding document](https://docs.strongswan.org/docs/5.9/howtos/forwarding.html#_hosts_on_the_lan)).

> The route is also visible and announced to standard-Routing Table. So you can see it also with netstat

Why would that be necessary? Seems redundant (and could potentially even cause conflicts - the whole point of installing routes in table 220 is to avoid those).

**Reply by pedda4711** [reply]
if you need to import the route (host or net) and distribute it to other hops it is necessary that you can export the routes to standard routing-tables so daemons as quagga or frr or bird can advertise this to their members.  Importet routes can exportet or not via filters in the daemons.

**Reply by tobiasbrunner** [reply]
Sorry, I'm not sure I understand. If a routing daemon imports the routes from a particular table, it could already announce them to other peers, no? Why would the routes have to be copied to the main routing table? Or is that just something a routing daemon does (i.e. it announces the routes also to the host on which it is running itself)?

**Reply by pedda4711** [reply]
The routing daemon import table 220 and also announce this routes to visible routes in standard table on the server self (filtered if needed). It also announce it to other hosts (filtered if needed). There is no routing-problem on the local server.  Standard-table show also connected routes.
I've a client (host or net) connect, the route is published to internal network behind a dmz and if needed to other connected networks

In my setup i have multiple vpn-servers  (for Users, for networks) with a physical dmz between Internet and local Network. 
I need transparent routing because some services does not work with nat. I've also "private ip" based firewalls in the internal network so users or networks with low rights don't have access to many servers in internal / connected private Network.  

with this setup one user can connect to VPN-Server A  or B or C, get a IP/32 from radius and the complete network know about this route.
Also a network can connect to VPN-Server A or B or C (Failover) and the rest of the network know the way
